### DISCUSSION TITLE
DefinePlugin with not replacing function

### DISCUSSION THREAD

**Discussion by khats** [webpack/webpack#14020] [Q&A]
I try to replace function declaration with DefinePlugin and it worked with 4th version of webpack, but 5th (5.26.2) not;
I have repository with demo:
https://github.com/khats/storybook-demo
and issue reported to storybook project https://github.com/storybookjs/storybook/issues/14257

In webpack config there is:
```
plugins: [
      new webpack.DefinePlugin({
      "process.env.NODE_ENV": JSON.stringify("development"),
      "window.FAKE_GLOBAL_FUNCTION": "(function() { return console.log('log from define plugin function'); })",
    }),
],
```

In code this declaration not replaced with DefinePlugin:
```
window.FAKE_GLOBAL_FUNCTION();
```
but this one:
```
const tempVariable = window.FAKE_GLOBAL_FUNCTION; 
tempVariable();
```
compiled to:
```
var tempVariable = (function() { return console.log('log from define plugin function'); });
tempVariable();
```

I have added my note to storybook ticket too: https://github.com/storybookjs/storybook/issues/14257#issuecomment-901233371

**Comment by alexander-akait** [comment]
It is described on docs, only identifiers replaced https://webpack.js.org/plugins/define-plugin/#usage, not functions

**Reply by khats** [reply]
@alexander-akait why this one:
```window.FAKE_GLOBAL_FUNCTION();```
not replaced with this one?
```(function() { return console.log('log from define plugin function'); })();```

**Reply by tophf** [reply]
@alexander-akait, we can do it in v5 anyway via `(0, process.env.FOO)()`, so why not enable proper substitution in `evaluateCallExpression` parser hook (or a similar one) by default?

**Comment by khats** [comment]
define variable as (without "window." prefix in variable) works too:
```
plugins: [
      new webpack.DefinePlugin({
      "process.env.NODE_ENV": JSON.stringify("development"),
      "FAKE_GLOBAL_FUNCTION": "(function() { return console.log('log from define plugin function'); })",
    }),
],

```
